File: src/main/java/me/m64diamondstar/ingeniamccore/Main.java
package me.m64diamondstar.ingeniamccore;

import me.m64diamondstar.ingeniamccore.Database.MySQL;
import me.m64diamondstar.ingeniamccore.General.Commands.GamemodeCmd;
import me.m64diamondstar.ingeniamccore.General.Commands.Ingenia.IngeniaCmd;
import me.m64diamondstar.ingeniamccore.Wands.WandListener.WandListener;
import org.bukkit.Bukkit;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.Objects;

public final class Main extends JavaPlugin {

    public MySQL SQL;

    @Override
    public void onEnable() {
        Bukkit.getLogger().info("---------------------------");
        Bukkit.getLogger().info("Started loading IngeniaMC-Core!");
        Bukkit.getLogger().info(" ");

        loadMainInstances();
        Bukkit.getLogger().info("Main instances loaded ✓");

        this.saveDefaultConfig();
        this.reloadConfig();
        Bukkit.getLogger().info("Config (re)loaded ✓");
        this.SQL = new MySQL(this);


        this.loadCommandExecutors();
        Bukkit.getLogger().info("Commands loaded ✓");

        this.loadTabCompleters();
        Bukkit.getLogger().info("Tab Completers loaded ✓");

        this.loadEventListeners();
        Bukkit.getLogger().info("Event Listeners loaded ✓");

        Bukkit.getLogger().info("Scoreboard Library loaded ✓");

        Bukkit.getLogger().info(" ");
        Bukkit.getLogger().info("Finished loading, IngeniaMC-Core is enabled!");
        Bukkit.getLogger().info("---------------------------");
    }



    @Override
    public void onDisable() {
        SQL.disconnect();
    }





    private void loadMainInstances(){

    }

    private void loadCommandExecutors(){
        Objects.requireNonNull(this.getCommand("gmc")).setExecutor(new GamemodeCmd());
        Objects.requireNonNull(this.getCommand("gms")).setExecutor(new GamemodeCmd());
        Objects.requireNonNull(this.getCommand("gma")).setExecutor(new GamemodeCmd());
        Objects.requireNonNull(this.getCommand("gmsp")).setExecutor(new GamemodeCmd());

        Objects.requireNonNull(this.getCommand("ingenia")).setExecutor(new IngeniaCmd());
    }

    private void loadTabCompleters(){

    }

    private void loadEventListeners(){
        Bukkit.getServer().getPluginManager().registerEvents(new WandListener(), this);
    }

}

File: src/main/java/me/m64diamondstar/ingeniamccore/Main.kt
package me.m64diamondstar.ingeniamccore

import com.comphenix.protocol.ProtocolLibrary
import com.comphenix.protocol.ProtocolManager
import me.m64diamondstar.ingeniamccore.attractions.listeners.PlayerInteractEntityListener
import me.m64diamondstar.ingeniamccore.general.commands.CosmeticCommand
import me.m64diamondstar.ingeniamccore.general.commands.GamemodeCommand
import me.m64diamondstar.ingeniamccore.general.commands.MessageCommand
import me.m64diamondstar.ingeniamccore.general.commands.ingenia.IngeniaCommand
import me.m64diamondstar.ingeniamccore.general.commands.WandCommand
import me.m64diamondstar.ingeniamccore.general.commands.tabcompleters.IngeniaTabCompleter
import me.m64diamondstar.ingeniamccore.general.listeners.*
import me.m64diamondstar.ingeniamccore.wands.wandlistener.WandListener
import me.m64diamondstar.ingeniamccore.general.listeners.helpers.BonemealListener
import me.m64diamondstar.ingeniamccore.general.player.IngeniaPlayer
import me.m64diamondstar.ingeniamccore.utils.gui.GuiListener
import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin
import java.util.*

class Main : JavaPlugin() {

    private lateinit var protocolManager: ProtocolManager

    companion object {
        lateinit var plugin: Main
        var isDisabling: Boolean = false
    }

    override fun onEnable() {

        plugin = this
        protocolManager = ProtocolLibrary.getProtocolManager()

        Bukkit.getLogger().info("---------------------------")
        Bukkit.getLogger().info("Started loading IngeniaMC-Core!")
        Bukkit.getLogger().info(" ")

        Bukkit.getLogger().info("Main instances loaded ✓")

        saveDefaultConfig()
        reloadConfig()
        Bukkit.getLogger().info("Config (re)loaded ✓")

        loadCommandExecutors()
        Bukkit.getLogger().info("Commands loaded ✓")

        loadTabCompleters()
        Bukkit.getLogger().info("Tab Completers loaded ✓")

        loadEventListeners()
        Bukkit.getLogger().info("Event Listeners loaded ✓")

        loadPacketListeners()
        Bukkit.getLogger().info("Packet Listeners loaded ✓")

        for(player in Bukkit.getOnlinePlayers()){
            val ingeniaPlayer = IngeniaPlayer(player)
            ingeniaPlayer.setScoreboard(true)
        }
        Bukkit.getLogger().info("Player Scoreboards loaded ✓")

        Bukkit.getLogger().info(" ")
        Bukkit.getLogger().info("Finished loading, IngeniaMC-Core is enabled!")
        Bukkit.getLogger().info("---------------------------")
    }

    override fun onDisable() {
        isDisabling = true
    }

    private fun loadCommandExecutors() {
        Objects.requireNonNull(getCommand("gmc"))?.setExecutor(GamemodeCommand())
        Objects.requireNonNull(getCommand("gms"))?.setExecutor(GamemodeCommand())
        Objects.requireNonNull(getCommand("gma"))?.setExecutor(GamemodeCommand())
        Objects.requireNonNull(getCommand("gmsp"))?.setExecutor(GamemodeCommand())

        Objects.requireNonNull(getCommand("ingenia"))?.setExecutor(IngeniaCommand())

        Objects.requireNonNull(getCommand("cosmetics"))?.setExecutor(CosmeticCommand())

        Objects.requireNonNull(getCommand("wand"))?.setExecutor(WandCommand())

        Objects.requireNonNull(getCommand("msg"))?.setExecutor(MessageCommand())
        Objects.requireNonNull(getCommand("tell"))?.setExecutor(MessageCommand())
        Objects.requireNonNull(getCommand("r"))?.setExecutor(MessageCommand())
        Objects.requireNonNull(getCommand("react"))?.setExecutor(MessageCommand())
    }

    private fun loadTabCompleters() {
        Objects.requireNonNull(getCommand("ig")?.setTabCompleter(IngeniaTabCompleter()))
    }

    private fun loadEventListeners() {
        /*
            All wand events
         */
        Bukkit.getServer().pluginManager.registerEvents(WandListener(), this)

        /*
            Join/leave events
         */
        Bukkit.getServer().pluginManager.registerEvents(JoinListener(), this)
        Bukkit.getServer().pluginManager.registerEvents(LeaveListener(), this)
        Bukkit.getServer().pluginManager.registerEvents(ServerListPingListener(), this)

        /*
            Chat events
         */
        Bukkit.getServer().pluginManager.registerEvents(ChatListener(), this)

        /*
            Inventory/GUI open events
         */
        Bukkit.getServer().pluginManager.registerEvents(InteractListener(), this)
        Bukkit.getServer().pluginManager.registerEvents(InventoryListener(), this)

        /*
            All GUI events
         */
        Bukkit.getServer().pluginManager.registerEvents(GuiListener(), this)

        /*
            Help Events
         */
        Bukkit.getServer().pluginManager.registerEvents(BonemealListener(), this)

        /*
            Attraction Events
         */
        Bukkit.getServer().pluginManager.registerEvents(PlayerInteractEntityListener(), this)
    }

    private fun loadPacketListeners(){

    }
}